File: lowestCommonAncestorinbst.py
#Time Complexity is O(n)   -- n being number of nodes in Tree
#Space Complexity is O(h) -- h being height of Tree
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
# Your code here along with comments explaining your approach
#recursive
def lowestCommonAncestor(root, p, q):
    if root is None:
        return
    if root.value == p.value or root.value == q.value :
        return root
    m = lowestCommonAncestor(root.left, p, q)
    n = lowestCommonAncestor(root.right, p, q)
    if m is None:
        return n
    if n is None:
        return m
    return root

File: java/LowestCommonAncestorInBST.java
// Time Complexity : O(H)
// Space Complexity : O(H)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    TreeNode result;

    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) return root;
        helper(root, p, q);
        return result;
    }

    private void helper(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null) return;
        if (p.val < root.val && q.val < root.val) {
            helper(root.left, p, q);
        } else if (p.val > root.val && q.val > root.val) {
            helper(root.right, p, q);
        } else {
            result = root;

        }
    }
}
Approach 2------------------------
// Time Complexity : O(H)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode(int x) { val = x; }
 * }
 */

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {

        if (root == null) {
            return null;
        }

        while (root != null) {

            if (root.val > p.val && root.val > q.val) {
                root = root.left;
            } else if (root.val < p.val && root.val < q.val) {
                root = root.right;
            } else {

                return root;
            }
        }
        return null;

    }
}